mantidproject/mantid
# Implement CompressEvents method without sorting events

### Description of work

A new method to compress events has been implemented that when density of events are high will be faster. This implementation is based on [`CompressEventAccumulator::CompressDense`](https://github.com/mantidproject/mantid/blob/main/Framework/DataHandling/src/CompressEventAccumulator.cpp#L309)

This will probably conflict with changes in #37881, so when one is merged the other will need to be updated.


#### Summary of work




Fixes [4842: CompressEvents works with unsorted events](https://ornlrse.clm.ibmcloud.com/ccm/resource/itemName/com.ibm.team.workitem.WorkItem/4842)



#### Further detail of work


### To test:

Use a file with a large number of events per spectra, or group spectra/align detectors/sum speactra.

Using the example file given in the issue and summing spectra, compare between setting `SortFirst` to `True` or `False`. I see it running about twice as fast when `SortFirst=False`.

```python
filename = "/SNS/SNAP/IPTS-28913/nexus/SNAP_57514.nxs.h5"
ws = LoadEventNexus(Filename=filename, NumberOfBins=1)
ws = SumSpectra(ws)
ws = CompressEvents(ws, 0.01, SortFirst=False)
```







---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
